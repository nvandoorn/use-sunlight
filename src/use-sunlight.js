import { useState } from 'react'
import { useGeolocation as globalUseGeolocation } from './use-geolocation'
import { useInterval } from './use-interval'
// TODO roll something smaller since
// this package likely contains tons
// of code we dont need
import suncalc from 'suncalc'

export const computeLightLevel = (
  { loading, latitude, longitude },
  date = new Date()
) => {
  // assume darkness if info is not ready/available
  if (loading || !latitude || !longitude) return 0
  let { altitude } = suncalc.getPosition(date, latitude, longitude)
  if (altitude < 0) {
    altitude += 2 * Math.PI
  }
  // on or below the horizon is considered "no light"
  if (altitude >= Math.PI) return 0
  // otherwise, return a relative altitude
  // where n is an int in [0,10]
  // and 10 represents the zenith (straight over your head)
  return Math.round(10 * Math.sin(altitude))
}

// make a location hook injectable such that we can test this
export const useSunlight = (useGeolocation = globalUseGeolocation) => {
  const geoLocation = useGeolocation()
  const getLightLevel = () => computeLightLevel(geoLocation)

  const [lightLevel, setLightLevel] = useState(getLightLevel())
  useInterval(() => {
    setLightLevel(getLightLevel())
  }, 1000)
  return [lightLevel]
}
