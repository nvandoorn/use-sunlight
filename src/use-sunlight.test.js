import { computeLightLevel } from './use-sunlight'
// TODO figure out why this is required
// (I'm assuming some transpiler magic
// makes this "required")
// eslint-disable-next-line
import React from 'react'

// Test the calculations used in the hook.
// Not sure if it's worth testing the public hook
// API
test('computeLightLevel', () => {
  expect(
    computeLightLevel({ latitude: 0, longitude: 0, loading: false })
  ).toMatchSnapshot()
})
