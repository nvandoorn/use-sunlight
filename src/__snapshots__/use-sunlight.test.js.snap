// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`computeLightLevel 1`] = `0`;
